File: R/derivative.R
##' .. content for \description{} (no empty lines) ..
##'
##' .. content for \details{} ..
##' @title Scharr image derivative
##' @param img rip object
##' @param dx order of the derivative in x direction.
##' @param dy order of the derivative n y direction.
##' @param scale optional scale factor for the computed derivative values. By default no scale is being applied.
##' @param delta optional delta value that is added to the results prior to output.
##' @return rip object of the same size and same number of channels as \code{img}.
##' @author Kaustav Nandy
Scharr <- function(img, dx, dy, scale=1, delta=0, bordertype=4)
{
    mderiv <- Module("derivative", PACKAGE = "RIPOpenCV")
    mderiv$scharr(img, dx, dy, scale, delta)
}


##' .. content for \description{} (no empty lines) ..
##'
##' .. content for \details{} ..
##' @title Derivative using Sobel operator
##' @param img rip object.
##' @param dx order of the derivative in x direction
##' @param dy order of the derivative in y direction
##' @param ksize size of the extended sobel kernel, it must be 1, 3, 5 or 7
##' @param scale optional scale factor for the computed derivative values. By default no scale is being applied.
##' @param delta optional delta value that is added to the results prior to output.
##' @return rip object of the same size and same number of channels as \code{img}.
##' @author Kaustav Nandy
Sobel <- function(img, dx, dy, ksize, scale=1, delta=0, bordertype=4)
{
    mderiv <- Module("derivative", PACKAGE = "RIPOpenCV")
    mderiv$sobel(img, dx, dy, ksize, scale, delta)
}

##' .. content for \description{} (no empty lines) ..
##'
##' .. content for \details{} ..
##' @title Applies seperable filter to an image
##' @param img rip object
##' @param kernelX Coefficients for filtering each row.
##' @param kernelY Coefficients for filtering each column.
##' @param anchor Anchor position within the kernel. The default value (|1,|1) means that the anchor is at the kernel center.
##' @param delta Value added to the filtered results before returning them.
##' @return rip object of same size and number of channel as \code{img}.
##' @author Kaustav Nandy
sepFilter2D <- function(img, kernelX, kernelY, anchor, delta, bordertype=4)
{
    mderiv <- Module("derivative", PACKAGE = "RIPOpenCV")
    mderiv$sepFilter(img, ddepth, kernelX, kernelY, anchor, delta)
}

File: src/derivative.cpp
#include "conversion.h"

using namespace cv;
using namespace Rcpp;

Rcpp::NumericMatrix scharr(Rcpp::NumericMatrix imgMat, int ddepth,
			   int dx, int dy, double scale, double delta)
{
    cv::Mat M = convertMat_RCPP2CV(imgMat, 5);
    cv::Mat outImg;
    cv::Scharr(M, outImg, ddepth, dx, dy, scale, delta);
    return (convertMat_CV2RCPP(outImg));
}



Rcpp::NumericMatrix sepFilter(Rcpp::NumericMatrix imgMat, int ddepth, 
			      Rcpp::NumericMatrix kernelX_, Rcpp::NumericMatrix kernelY_, 
			      Rcpp::NumericVector anchor_, double delta)
{
    cv::Mat M = convertMat_RCPP2CV(imgMat, 5);
    cv::Mat outImg;
    cv::Mat kernelX = convertMat_RCPP2CV(kernelX_, 5);
    cv::Mat kernelY = convertMat_RCPP2CV(kernelY_, 5);
    std::vector<int> anchor = as< std::vector<int> >(anchor_);
    cv::sepFilter2D(M, outImg, ddepth, kernelX, kernelY, 
		    Point(anchor.at(0), anchor.at(1)), delta);
    return (convertMat_CV2RCPP(outImg));
}


Rcpp::NumericMatrix sobel(Rcpp::NumericMatrix imgMat, int ddepth, int dx, int dy, 
			  int ksize, double scale, double delta)
{
    cv::Mat M = convertMat_RCPP2CV(imgMat, 5);
    cv::Mat outImg;
    cv::Sobel(M, outImg, ddepth, dx, dy, ksize, scale, delta);
    return (convertMat_CV2RCPP(outImg));
}

RCPP_MODULE(derivative) 
{
    function("scharr", &scharr, "Scharr derivative");
    function("sepFilter", &sepFilter, "Separate Filter to each direction");
    function("sobel", &sobel, "Sobel derivative");
}
